source/common/event/dispatcher_impl.cc

```diff
@@ -117,6 +121,71 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
                                                          std::move(transport_socket), options);
 }
 
+namespace {
+Network::Address::InstanceConstSharedPtr
+nextClientAddress(const Network::Address::InstanceConstSharedPtr& server_address) {
+  uint64_t id = 0;
+  return std::make_shared<Network::Address::EnvoyInternalInstance>(absl::StrCat(server_address->asStringView(), "_", ++id));
+}
+} // namespace
+
+Network::ClientConnectionPtr
+DispatcherImpl::createInternalConnection(Network::Address::InstanceConstSharedPtr internal_address,
+                                         Network::Address::InstanceConstSharedPtr local_address) {
+  ASSERT(isThreadSafe());
+  if (internal_address == nullptr) {
+    return nullptr;
+  }
+  if (local_address == nullptr) {
+    local_address = nextClientAddress(internal_address);
+  }
+  // Find the internal listener callback. The listener will setup the server connection.
+  auto iter = internal_listeners_.find(internal_address->asString());
+  for (const auto& [name, _] : internal_listeners_) {
+    ENVOY_LOG_MISC(debug, "lambdai: p listener {}", name);
+  }
+  if (iter == internal_listeners_.end()) {
```

Comment:
hm, I would have expected this in connect().  When we do normal IP style create and connect do we validate before connect?
no strong opinion either way, but if you prefer this maybe comment about where validation is done in include/ ?


Actually I think if you move this work to CONNECT you can skip the entire closingClientConnectionImpl which would be nice clean up

Reply:
The reason I want to split `ClosingClientConnectionImpl` and `ClientConnectionImpl`:

The underlying `BufferedIoSocketHandleImpl` is aware of single transition: from "connected" to "disconnected". 
If we set up the server connection in `connect()`, the client connection need to track "uninitialized" to "connected", "uninitialized" to "fail to connect".

BTW, The model internal connection I am borrowing from "socketpair()"